sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/batch/BulkExecutor.java

```diff
@@ -156,111 +159,172 @@ public BulkExecutor(CosmosAsyncContainer container,
+                                    logger.warn("Failed to cancel scheduled tasks{}", getThreadInfo(), e);
+                                }
+                            }
+
+                            this.onFlush();
+
+                            long flushIntervalAfterDrainingIncomingFlux = Math.min(
+                                this.maxMicroBatchIntervalInMs,
+                                BatchRequestResponseConstants
+                                    .DEFAULT_MAX_MICRO_BATCH_INTERVAL_AFTER_DRAINING_INCOMING_FLUX_IN_MILLISECONDS);
+
+                            this.scheduledFutureForFlush = this.executorService.scheduleWithFixedDelay(
+                                this::onFlush,
+                                flushIntervalAfterDrainingIncomingFlux,
+                                flushIntervalAfterDrainingIncomingFlux,
+                                TimeUnit.MILLISECONDS);
+                        }
+                    })
+                    .mergeWith(mainSink.asFlux())
+                    .subscribeOn(CosmosSchedulers.BULK_EXECUTOR_BOUNDED_ELASTIC)
+                    .flatMap(
+                        operation -> {
+                            logger.trace("Before Resolve PkRangeId, {}, Context: {} {}",
+                                getItemOperationDiagnostics(operation),
+                                this.operationContextText,
+                                getThreadInfo());
+
+                            // resolve partition key range id again for operations which comes in main sink due to gone retry.
+                            return BulkExecutorUtil.resolvePartitionKeyRangeId(this.docClientWrapper, this.container, operation)
+                                                   .map((String pkRangeId) -> {
+                                                       PartitionScopeThresholds partitionScopeThresholds =
+                                                           this.partitionScopeThresholds.computeIfAbsent(
+                                                               pkRangeId,
+                                                               (newPkRangeId) -> new PartitionScopeThresholds(newPkRangeId, this.cosmosBulkExecutionOptions));
+
+                                                       logger.trace("Resolved PkRangeId, {}, PKRangeId: {} Context: {} {}",
+                                                           getItemOperationDiagnostics(operation),
+                                                           pkRangeId,
+                                                           this.operationContextText,
+                                                           getThreadInfo());
+
+                                                       return Pair.of(partitionScopeThresholds, operation);
+                                                   });
+                        })
+                    .groupBy(Pair::getKey, Pair::getValue)
```

Comment:
Do we have any unit tests for BulkExecutor in Java? Could a UT trying to add more than 256 items where each item is for a different PKRangeId trigger the original bug?

Reply:
Yes - should be possible - but only helps if you have a container with >256 physical partitions - which isn't possible witht he emulator and would be extremely expensive during unit tests with real accounts (you would have to provision 1540000 RU to force > 256 partitions) and pay for one hour. So not planning to add this as a unit test.